source/extensions/dynamic_modules/sdk/rust/src/lib.rs

```diff
@@ -330,6 +330,22 @@ pub trait EnvoyHttpFilter {
   ///
   /// Returns true if the operation is successful.
   fn set_response_trailer(&mut self, key: &str, value: &[u8]) -> bool;
+
+  /// Get the number-typed dynamic metadata value with the given key.
+  /// If the metadata is not found, this returns `None`.
+  fn get_dynamic_metadata_number(&self, namespace: &str, key: &str) -> Option<f64>;
+
+  /// Set the number-typed dynamic metadata value with the given key.
+  /// Returns true if the operation is successful.
+  fn set_dynamic_metadata_number(&mut self, namespace: &str, key: &str, value: f64) -> bool;
+
+  /// Get the string-typed dynamic metadata value with the given key.
+  /// If the metadata is not found, this returns `None`.
+  fn get_dynamic_metadata_string(&self, namespace: &str, key: &str) -> Option<EnvoyBuffer>;
```

Comment:
In looking at this again I'm realizing that there are some fairly easy ways to create unsound code with the way EnvoyBuffer works. Let's say you return an EnvoyBuffer, then delete the underlying data (header/metadata). This will be unsound. 

I haven't thought about it in detail but I think the returned EnvoyBuffer has to have lifetime mapped to the underlying EnvoyFilter. This should prevent issues like this. Multiple &self borrows should be ok but will be blocked by the &mut self borrow required for mutation.

Let's try to figure this out now before going further.

Reply:
yeah totally agree and easy to write an unsound code with this. I think maybe the easiest way is to mark everything returning EnvoyBuffer (ie non-copy versions) as unsafe (in Rust keyword level) and add safer methods which copy all the strings in the Rust side. But I guess there should be a way to do this just by relying on Rust's borrow checker etc. 

Let me think about it ....

Reply:
The solution as I said is to have the lifetime of envoybuffer bound to the underlying filter. More unsafe is not the solution. We should not ask any filter writer to use unsafe.

Reply:
I opened the separate PR: https://github.com/envoyproxy/envoy/pull/37880  it turns out relatively easy to fix this. Thank you for your suggestion!